### PR TITLE
Vault rounding and complementary debt functions 

### PR DESCRIPTION
**Pull request by DaigaroCota**
This pull request addresses multiple vulnerabilities and issues reported in the Macro report audit.

M-2 Vault rounding inconsistencies
L-6 Lack of option of fully payback loan
Q-2 Unnecessary use of _msgSender()
Q-3 Double conversion of assets<>shares in withdraw function flow
G-1 and G2 Immutable definition of vault asset and debtAsset

### CODE REVIEW

File: packages/protocol/src/interfaces/IVault.sol [hunk 1]
```diff
@@ -213,7 +213,48 @@ interface IVault is IERC4626 {
...
+   * @param owner to check
+   *
+   * @dev Requirements:
+   * - Must not revert.
+   */
+  function maxPayback(address owner) external view returns (uint256 debt);
+
+  /**
+   * TODO
+   */
```
Comment (pedrovalido) [inline packages/protocol/src/interfaces/IVault.sol:231]: Dont forget this //TODO

File: packages/protocol/src/interfaces/IVault.sol [hunk 2]
```diff
@@ -213,7 +213,48 @@ interface IVault is IERC4626 {
...
+  function maxMintDebt(address owner) external view returns (uint256 shares);
+
+  /**
+   * TODO
+   */
```
Comment (pedrovalido) [inline packages/protocol/src/interfaces/IVault.sol:236]: Dont forget this //TODO

File: packages/protocol/src/interfaces/IVault.sol [hunk 3]
```diff
@@ -213,7 +213,48 @@ interface IVault is IERC4626 {
...
+  function maxBurnDebt(address owner) external view returns (uint256 shares);
+
+  /**
+   * TODO
+   */
```
Comment (pedrovalido) [inline packages/protocol/src/interfaces/IVault.sol:241]: Dont forget this //TODO

File: packages/protocol/src/interfaces/IVault.sol [hunk 4]
```diff
@@ -213,7 +213,48 @@ interface IVault is IERC4626 {
...
+  function previewBorrow(uint256 debt) external view returns (uint256 shares);
+
+  /**
+   * TODO
```
Comment (pedrovalido) [inline packages/protocol/src/interfaces/IVault.sol:245]: Dont forget this //TODO

File: packages/protocol/src/interfaces/IVault.sol [hunk 5]
```diff
@@ -213,7 +213,48 @@ interface IVault is IERC4626 {
...
+   */
+  function previewMintDebt(uint256 shares) external view returns (uint256 debt);
+
+  /**
+   * TODO
```
Comment (pedrovalido) [inline packages/protocol/src/interfaces/IVault.sol:250]: Dont forget this //TODO

File: packages/protocol/src/interfaces/IVault.sol [hunk 6]
```diff
@@ -213,7 +213,48 @@ interface IVault is IERC4626 {
...
+   */
+  function previewPayback(uint256 debt) external view returns (uint256 shares);
+
+  /**
+   * TODO
```
Comment (pedrovalido) [inline packages/protocol/src/interfaces/IVault.sol:255]: Dont forget this //TODO

File: packages/protocol/src/interfaces/IVault.sol [hunk 7]
```diff
@@ -228,7 +269,18 @@ interface IVault is IERC4626 {
    * - Must revert if owner does not own sufficient collateral to back debt.
    * - Must revert if caller is not owner or permissioned operator to act on owner behalf.
    */
-  function borrow(uint256 debt, address receiver, address owner) external returns (uint256);
+  function borrow(uint256 debt, address receiver, address owner) external returns (uint256 shares);
+
+  /**
+   * TODO
```
Comment (pedrovalido) [inline packages/protocol/src/interfaces/IVault.sol:275]: Dont forget this //TODO

File: packages/protocol/src/interfaces/IVault.sol [hunk 8]
```diff
@@ -240,7 +292,12 @@ interface IVault is IERC4626 {
    * Requirements:
    * - Must emit a Payback event.
    */
-  function payback(uint256 debt, address receiver) external returns (uint256);
+  function payback(uint256 debt, address receiver) external returns (uint256 shares);
+
+  /**
+   * TODO
+   */
```
Comment (pedrovalido) [inline packages/protocol/src/interfaces/IVault.sol:299]: Dont forget this //TODO

**Review by pedrovalido — CHANGES_REQUESTED** [review]
Good job overall :)

**Comment by DaigaroCota** [comment]
@pedrovalido updated all the missing docs in TODOs. This should be ready to be merged.

**Comment by pedrovalido** [comment]
Great! Let's merge then

**Pull request merged**